src/docs/arch-compact.dox

```diff
@@ -0,0 +1,57 @@
+the @ref arch-block "Block Manager" to reduce the on-disk footprint of WT tables. The compaction
+process is initiated by user application by calling \c WT_SESSION::compact method. Overall, the
+compaction process involves rewriting disk blocks from the end of the on-disk file to the start of
+the file, hence giving an opportunity to reclaim the space from the end of file through truncation
+and effectively reducing the file system reported file size. Note that the compaction is a
+"best effort" process and does not guarantees reduction in file size.
+
+Before understanding compaction process, we need to understand how disk blocks are referenced by the
+checkpoints. Each checkpoint references a certain set of disk blocks for a table.. When a dirty
+btree page is reconciled, a new disk block is assigned for the page and a new checkpoint starts
+referencing the new block used for the page. Once the checkpoint referencing the old block is deleted,
+the old disk block becomes available for reuse. The aim of compaction is to copy over the data from
+the blocks at the end of file to the blocks in the start of the file that are available for reuse.
+
+The compaction process starts with a system-wide checkpoint. There are potentially many dirty blocks
+in the cache, and the intention here is to write them out and then discard previous checkpoints so
+that there are as many blocks as possible on the file's "available for reuse" list when the compaction
+starts.  The compaction logic then starts examining the files that need to be compacted. Compacting
+the object is done 10% at a time, that is, compaction tries to move blocks from the last 10% of the
+file into the beginning of the file (the 10% is hard coded in the block manager).  The reason for
+this is because compaction walks the file in logical order, not block offset order, and compaction
+of a file can fail if the block from the end of the file is not written first. Note that the block manager
+uses a first-fit block selection algorithm during compaction to maximize block movement. The process
+is repeated multiple times until block manager detects that there no blocks from the last 10% of the file
+that can be moved.
+
+Diving deep into the compaction process, the btree layer walks the internal pages for a given table to
+identify leaf pages that can be rewritten. For each leaf page referenced by the internal page, we use
+a block manager API to assess whether it will be beneficial to rewrite the leaf page. If the block
+manager recognizes that rewriting the page will be beneficial, it rewrites the disk
+block to a new disk location. The new disk address is placed into the parent internal page and
+the page is marked dirty. In this way the leaf page is rewritten without being brought into the
+cache. If the leaf page is already in cache, the leaf page is simply marked dirty and the subsequent
+checkpoint or eviction rewrites the page.
```

Comment:
We need to mention that compaction moves internal pages too. And the root page will be rewritten by the checkpoint. You can find comments about that in `__compact_walk_internal`.

Reply:
I discuss this more in ce28d39.